junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineDiscoveryResultValidator.java

```diff
@@ -37,30 +38,39 @@ void validate(TestEngine testEngine, TestDescriptor root) {
 			() -> String.format(
 				"The discover() method for TestEngine with ID '%s' must return a non-null root TestDescriptor.",
 				testEngine.getId()));
-		Preconditions.condition(isAcyclic(root),
-			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph.",
-				testEngine.getId()));
+		Optional<UniqueId> cyclicUIDs = getFirstCyclicUID(root);
+		Preconditions.condition(!cyclicUIDs.isPresent(),
+			() -> String.format("The discover() method for TestEngine with ID '%s' returned a cyclic graph: "
+					+ "Unique ID(s) '%s' enter a cycle",
+				testEngine.getId(), cyclicUIDs));
 	}
 
 	/**
 	 * @return {@code true} if the tree does <em>not</em> contain a cycle; else {@code false}.
 	 */
 	boolean isAcyclic(TestDescriptor root) {
+		return !getFirstCyclicUID(root).isPresent();
```

Comment:
Is this method still used by non-test code?

Reply:
No, and after the latest change it has now become getCyclicGraphInfo, which contains a short visualisation of the cycle:

```
				"Test [engine:root] exists in at least two paths:\n" +
						"\t[engine:root]\n" +
						"\t[engine:root] <- [engine:root]"
```